src/snabbdom.ts

```diff
@@ -60,125 +95,111 @@ export function init(modules: Array<Partial<Module>>, domApi?: DOMAPI) {
-            api.appendChild(elm, createElm(ch as VNode, insertedVnodeQueue));
+          if (isVNode(ch)) {
+            api.appendChild(elm, createElm(ch, insertedVnodeQueue));
           }
         }
-      } else if (is.primitive(vnode.text)) {
+      } else if (!isUndef(vnode.text)) {
         api.appendChild(elm, api.createTextNode(vnode.text));
       }
-      i = (vnode.data as VNodeData).hook; // Reuse variable
-      if (isDef(i)) {
-        if (i.create) i.create(emptyNode, vnode);
-        if (i.insert) insertedVnodeQueue.push(vnode);
-      }
+      if (isDef(i = data.hook) && isDef(i.insert)) insertedVnodeQueue.push(vnode);
+    } else if (isTextVNode(vnode)) {
+      vnode.elm = api.createTextNode(vnode.text);
+    } else if (isCommentVNode(vnode)) {
+      vnode.elm = api.createComment(vnode.text);
+    } else if (isFragmentVNode(vnode)) {
+      vnode.elm = api.createFragment();
     } else {
-      vnode.elm = api.createTextNode(vnode.text as string);
+      throw new TypeError();
     }
     return vnode.elm;
   }
 
   function addVnodes(parentElm: Node,
                      before: Node | null,
-                     vnodes: Array<VNode>,
+                     vnodes: VNodes,
                      startIdx: number,
                      endIdx: number,
                      insertedVnodeQueue: VNodeQueue) {
     for (; startIdx <= endIdx; ++startIdx) {
       const ch = vnodes[startIdx];
-      if (ch != null) {
+      if (isVNode(ch)) {
         api.insertBefore(parentElm, createElm(ch, insertedVnodeQueue), before);
       }
     }
   }
 
   function invokeDestroyHook(vnode: VNode) {
     let i: any, j: number, data = vnode.data;
-    if (data !== undefined) {
-      if (isDef(i = data.hook) && isDef(i = i.destroy)) i(vnode);
-      for (i = 0; i < cbs.destroy.length; ++i) cbs.destroy[i](vnode);
-      if (vnode.children !== undefined) {
-        for (j = 0; j < vnode.children.length; ++j) {
-          i = vnode.children[j];
-          if (i != null && typeof i !== "string") {
-            invokeDestroyHook(i);
-          }
-        }
+    if (isDef(i = data.hook) && isDef(i = i.destroy)) i(vnode);
+    for (i = 0; i < cbs.destroy.length; ++i) cbs.destroy[i](vnode);
+    const { children } = vnode;
+    if (isUndef(children)) {
+      return;
+    }
+    for (j = 0; j < children.length; ++j) {
+      i = children[j];
+      if (!isTextVNode(i)) {
+        invokeDestroyHook(i);
       }
     }
   }
 
   function removeVnodes(parentElm: Node,
-                        vnodes: Array<VNode>,
+                        vnodes: VNodes,
                         startIdx: number,
                         endIdx: number): void {
     for (; startIdx <= endIdx; ++startIdx) {
       let i: any, listeners: number, rm: () => void, ch = vnodes[startIdx];
-      if (ch != null) {
-        if (isDef(ch.sel)) {
+      // text nodes do not have logic associated to them
+      if (isVNode(ch)) {
+        if (!isTextVNode(ch)) {
```

Comment:
this still requires more thoughts in my opinion, why are text nodes special? what about comments, fragments, etc? should they also be skipped the destroy routine?